feat(core): Resources api change

Humberd opened the pull request:
* fix(core): getting resource value throws an error instead of returning undefined
* fix(core): narrow error type for resources API
* fix(core): move reload method from Resource to WritableResource
* build: use es2022 lib

Humberd committed:
fix(core): getting resource value throws an error instead of returning undefined

When there is an underlying error state it would not be possible to swallow the error with:
`computed(() => res.value()?.inner);`

BREAKING CHANGE:
* `Resource.value()` now throws an error when it's in an error state. Previously it returned `undefined`.

Humberd committed:
fix(core): narrow error type for resources API

`Resource.error` used to return `unknown`. Now it's `Error | undefined`.
For non-`Error` types they are encapsulated with the `Error` type.

BREAKING CHANGE:
* `Resource.error` now has a type of `Signal<Error | undefined>` instead of `Signal<unknown>`

Humberd committed:
fix(core): move reload method from Resource to WritableResource

Now only mutable resources can be reloaded

BREAKING CHANGE:
* `Resource.reload` has been moved to `WritableResource.reload`.

Humberd committed:
build: use es2022 lib

It was set to es2022 for build tsconfig in https://github.com/angular/angular/pull/60066/files#diff-7ec658aedc62a693cc513accc8f360938d90bcbdb09aa79c1401e56fe84f5fd3, but was missed for the main one.

JeanMeche commented on packages/common/http/src/resource.ts:
```diff
@@ -29,6 +29,7 @@ import {HttpEventType, HttpProgressEvent, HttpResponseBase} from './response';
 import {HttpHeaders} from './headers';
 import {HttpParams} from './params';
 import {HttpResourceRef, HttpResourceOptions, HttpResourceRequest} from './resource_api';
+import {encapsulateResourceError} from '../../../core/src/resource/resource';
```
```suggestion
import {encapsulateResourceError} from '@angular/core';
```
This is an external entry, the import shouldn't be relative

Humberd replied:
This function was very simple that I declared it independently in this file resulting in  no need to import it.

josephperrott commented on packages/tsconfig.json:
```diff
@@ -8,7 +8,7 @@
     "emitDecoratorMetadata": true,
     "module": "esnext",
-    "target": "es2020",
+    "target": "es2022",
```
Additionally, I am not sure what the reason this bump was needed based on the changes being made in the PR, though I admittedly don't know the exact changes from `es2020` to `es2022` to know how that plays here.

JeanMeche replied:
This change shouldn't be needed, only the `libs` to support the new `Error` signature.

Humberd replied:
You are right. Fixed.

Humberd commented on packages/tsconfig.json:
```diff
@@ -29,7 +29,7 @@
     "rootDir": ".",
     "inlineSourceMap": true,
-    "lib": ["es2020", "dom", "dom.iterable"],
+    "lib": ["es2020", "es2022", "dom", "dom.iterable"],
```
Fixed.

JeanMeche commented on packages/common/http/src/resource.ts:
```diff
@@ -413,9 +413,12 @@ class HttpResponseResource implements Resource<HttpResponseBase | undefined> {
-  reload(): boolean {
-    // TODO: should you be able to reload this way?
-    return this.parent.reload();
+function encapsulateResourceError(error: unknown): Error {
+  if (error instanceof Error) {
+    return error;
   }
+
+  return new Error('Unknown error', {cause: error});
 }
```
Idealy we should do a private export from core to prevent this duplicate function. 

* export as `ɵencapsulateResourceError` in `packages/core/src/core_private_export.ts`
* and import `ɵencapsulateResourceError as encapsulateResourceError` here

Humberd replied:
Thank you! Done.

devversion commented on packages/tsconfig.json:
```diff
@@ -29,7 +29,7 @@
     "rootDir": ".",
     "inlineSourceMap": true,
-    "lib": ["es2020", "dom", "dom.iterable"],
+    "lib": ["es2022", "dom", "dom.iterable"],
```
FYI that this change is fine because the Bazel tsconfig already uses `es2022` too: https://github.com/angular/angular/blob/7825c0419f703118a8ef57f07e201b641b7a2cda/packages/tsconfig-build.json#L22

josephperrott reviewed (approved):
LGTM

Humberd commented:
After the approvals were done I changed the error message from: 'Resource failed to load' to `Resource is currently in the error state'.

I also removed untracking of the stream https://github.com/angular/angular/pull/60610/files#diff-16f6b7cc4c4e69b511c8312d38c6cf71c412e6f1d6ac9092054c0f349215cdedR418, because there were some cases in g3, where `hasValue` was out of sync with `value` signal and in result was throwing errors.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

devversion reviewed (approved):
LGTM

Reviewed-for: public-api

Humberd committed:
fix(core): reading resource value after reload in the error state

When the resource is loading after reloading from the error state reading `Resource.value()` would return the default value instead of throwing an error.
This change prevents `Resource.hasValue()` from throwing an error in such a case.

BREAKING CHANGE:
* `Resource.value()` now returns a default value when in a loading state after reloading the error state

crisbeto reviewed (approved):
Reviewed-for: public-api

JeanMeche commented:
Changes landed as part of #61441
